lib/persisted-model.js

```diff
@@ -839,21 +840,64 @@ PersistedModel.replicate = function(since, targetModel, options, callback) {
   options = options || {};
 
   var sourceModel = this;
-  var diff;
-  var updates;
-  var Change = this.getChangeModel();
+  callback = callback || function defaultReplicationCallback(err) {
+    if (err) throw err;
+  };
+
+  debug('replicating %s since %s to %s since %s',
+    sourceModel.modelName,
+    since.source,
+    targetModel.modelName,
+    since.target);
+  if (options.filter) {
+    debug('\twith filter %j', options.filter);
+  }
+
+  // In order to avoid a race condition between the replication and
+  // other clients modifying the data, we must create the new target
+  // checkpoint as the first step of the replication process.
+  // As a side-effect of that, the replicated changes are associated
+  // with the new target checkpoint. This is actually desired behaviour,
+  // because that way clients replicating *from* the target model
+  // since the new checkpoint will pick these changes up.
+  // However, it increases the likelihood of (false) conflicts being detected.
+  // In order to prevent that, we run the replication multiple times,
+  // until no changes were replicated, but at most MAX_ATTEMPTS times
+  // to prevent starvation. In most cases, the second run will find no changes
+  // to replicate and we are done.
+  var MAX_ATTEMPTS = 3;
+
+  run(1, since);
+
+  function run(attempt, since) {
+    debug('\titeration #%s', attempt);
+    tryReplicate(sourceModel, targetModel, since, options, next);
+
+    function next(err, conflicts, cps, updates) {
+      var finished = err || conflicts.length ||
```

Comment:
assign some of the sub results to descriptive variables... but I can live with this